### ISSUE TITLE
Docker CE 24.0.5's sysvinit script does not mount a missing `/sys/fs/cgroup`

### ISSUE THREAD

**Issue by jrshaffe** [docker/cli#4462]
### Description

I have a container with Ubuntu 22.04.2 LTS and docker-ce 24.05 installed on it.  I am unable to start the docker service.  If I uninstall  docker-ce 24.05 and then install version `5:24.0.4-1~ubuntu.22.04~jammy` I can start docker

### Reproduce

Try and run `service docker start` 

### Expected behavior

Docker starts

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.6
 Git commit:        ced0996
 Built:             Fri Jul 21 20:35:18 2023
 OS/Arch:           linux/amd64
 Context:           default
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

### docker info

```bash
Client: Docker Engine - Community
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.6
 Git commit:        ced0996
 Built:             Fri Jul 21 20:35:18 2023
 OS/Arch:           linux/amd64
 Context:           default
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
root@4e495ee2-42f2-43cf-795d-816af3152bc0:/tmp/build/80754af9# docker info
Client: Docker Engine - Community
 Version:    24.0.5
 Context:    default
 Debug Mode: false

Server:
ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

### Additional Info

looking in /var/log/docker.log I see one error 

time="2023-07-25T22:18:23.472291174Z" level=error msg="failed to initialize a tracing processor \"otlp\"" error="no OpenTelemetry endpoint: skip plugin"

and then this line at the end of the output:
> failed to start daemon: Devices cgroup isn't mounted

**Comment by neersighted** [comment]
Hey there,

This is something of an intentional change -- previously, `/etc/init.d/docker` would mount the `cgroupfs` filesystem if it was missing. This was a legacy behavior dating back nearly a decade, and was finally removed in https://github.com/moby/moby/pull/46044 as it is unwanted on almost every modern system.

To give better advice, you likely will have to share more specifics about your container setup; however, I can give some generalized advice:
- If you're running Docker-in-Docker, there's some additional logic needed; the easiest way to do so is the `docker:24.0.5-dind` and similar images.
- If you're not running a Docker container (e.g. LXC, LXD), you should consider switching to systemd as your init system, which handles all this for you and is much better supported/tested.
- If the above bullet point is true, but you must use sysvinit, install the `cgroupfs-mount` package, which contains the correct logic to mount this filesystem. This package is `Suggests:` for `docker-ce`.
- If you're building your own Docker-in-Docker image from a full Ubuntu base image, and starting `dockerd` with the init script (again, you really shouldn't prefer to do this), install `cgroupfs-mount` and `service cgroupfs-mount start` before you `service docker start`.

In short: it sounds like you have a relatively exotic configuration that was relying on a really old hack. You should first prefer to use a less exotic configuration, or if that is not possible, use the cgroupfs logic provided by the operating system (https://github.com/tianon/cgroupfs-mount) over the inferior copy that used to be embedded in the packages.

**Comment by vvoland** [comment]
Let me close this issue for now. Feel free to continue the conversation though!

**Comment by jrshaffe** [comment]
@neersighted thanks for you quick response!  We are basing our image from a full Ubuntu base image.  Installing the `cgroupfs-mount` and starting it before starting docker got us past that issue for now.  We'll take a look at some of the other options you described when we get a chance.  Thanks again for your help!!
